Index: mt76x2e-p4rev-113843/src/ap/ap_cfg.c
===================================================================
--- mt76x2e-p4rev-113843.orig/src/ap/ap_cfg.c
+++ mt76x2e-p4rev-113843/src/ap/ap_cfg.c
@@ -7363,7 +7363,7 @@ INT Set_AutoChannelSelCheckTime_Proc(
 
 INT Show_DriverInfo_Proc(RTMP_ADAPTER *pAd, PSTRING arg)
 {
-	DBGPRINT(RT_DEBUG_OFF, ("driver version: %s (%s %s) .\n", AP_DRIVER_VERSION, __DATE__, __TIME__));
+	DBGPRINT(RT_DEBUG_OFF, ("driver version: %s .\n", AP_DRIVER_VERSION));
 
 #ifdef CONFIG_ANDES_SUPPORT
 	if (pAd->chipCap.MCUType == ANDES) { 
Index: mt76x2e-p4rev-113843/src/common/cmm_info.c
===================================================================
--- mt76x2e-p4rev-113843.orig/src/common/cmm_info.c
+++ mt76x2e-p4rev-113843/src/common/cmm_info.c
@@ -41,7 +41,7 @@ INT Set_DriverVersion_Proc(
 {
 #ifdef CONFIG_AP_SUPPORT
 	IF_DEV_CONFIG_OPMODE_ON_AP(pAd)
-		DBGPRINT(RT_DEBUG_OFF, ("Driver version-%s %s %s\n", AP_DRIVER_VERSION, __DATE__, __TIME__));
+		DBGPRINT(RT_DEBUG_OFF, ("Driver version-%s\n", AP_DRIVER_VERSION));
 #endif /* CONFIG_AP_SUPPORT */
 
 
