Index: mt76x2e-p4rev-113843/src/os/linux/config.mk
===================================================================
--- mt76x2e-p4rev-113843.orig/src/os/linux/config.mk
+++ mt76x2e-p4rev-113843/src/os/linux/config.mk
@@ -240,6 +240,8 @@ LD := $(CROSS_COMPILE)ld
 
 WFLAGS := -g -DAGGREGATION_SUPPORT -DPIGGYBACK_SUPPORT -DWMM_SUPPORT  -DLINUX -Wall -Wstrict-prototypes -Wno-trigraphs -DENHANCED_STAT_DISPLAY
 WFLAGS += -DSYSTEM_LOG_SUPPORT -DRT28xx_MODE=$(RT28xx_MODE) -DCHIPSET=$(MODULE) -DRESOURCE_PRE_ALLOC -DDBG_DIAGNOSE -DDBG_RX_MCS -DDBG_TX_MCS
+WFLAGS += -Wno-error=incompatible-pointer-types
+
 #APsoc Specific
 WFLAGS += -DCONFIG_RA_NAT_NONE
 #end of /* APsoc Specific */
