Index: mt76x2e-p4rev-113843/src/os/linux/cfg80211/cfg80211_util.c
===================================================================
--- mt76x2e-p4rev-113843.orig/src/os/linux/cfg80211/cfg80211_util.c
+++ mt76x2e-p4rev-113843/src/os/linux/cfg80211/cfg80211_util.c
@@ -28,6 +28,8 @@
 #include "rtmp_comm.h"
 #include "rtmp_osabl.h"
 #include "rt_os_util.h"
+#include "linux/nl80211.h"
+#include "linux/ieee80211.h"
 
 /* all available channels */
 UCHAR Cfg80211_Chan[] = {
@@ -949,6 +951,14 @@ void CFG80211OS_P2pClientConnectResultIn
 
 BOOLEAN CFG80211OS_RxMgmt(IN PNET_DEV pNetDev, IN INT32 freq, IN PUCHAR frame, IN UINT32 len) 
 {
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,18,0))
+        return cfg80211_rx_mgmt(pNetDev->ieee80211_ptr,
+                                freq,
+                                0,       //CFG_TODO return 0 in dbm
+                                frame,
+                                len,
+                                NL80211_RXMGMT_FLAG_ANSWERED);
+#else
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,12,0))
 	return cfg80211_rx_mgmt(pNetDev->ieee80211_ptr,
                                 freq,
@@ -987,7 +997,7 @@ BOOLEAN CFG80211OS_RxMgmt(IN PNET_DEV pN
 #endif /* LINUX_VERSION_CODE: 3.4.0 */
 #endif /* LINUX_VERSION_CODE: 3.6.0 */
 #endif /* LINUX_VERSION_CODE: 3.12.0 */
-
+#endif /* LINUX_VERSION_CODE: 3.18.0 */
 }
 
 VOID CFG80211OS_TxStatus(IN PNET_DEV pNetDev, IN INT32 cookie, IN PUCHAR frame, IN UINT32 len, IN BOOLEAN ack)
@@ -1076,4 +1086,4 @@ VOID CFG80211OS_RecvObssBeacon(VOID *pCB
 #endif /* LINUX_VERSION_CODE: 3.8.0 */
 }
 
-#endif /* RT_CFG80211_SUPPORT */
\ No newline at end of file
+#endif /* RT_CFG80211_SUPPORT */
Index: mt76x2e-p4rev-113843/src/include/cfg80211.h
===================================================================
--- mt76x2e-p4rev-113843.orig/src/include/cfg80211.h
+++ mt76x2e-p4rev-113843/src/include/cfg80211.h
@@ -36,7 +36,11 @@ typedef enum _NDIS_HOSTAPD_STATUS {
 typedef struct __CFG80211_CB {
 
 	/* we can change channel/rate information on the fly so we backup them */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,7,0)
+	struct ieee80211_supported_band Cfg80211_bands[NUM_NL80211_BANDS];
+#else
 	struct ieee80211_supported_band Cfg80211_bands[IEEE80211_NUM_BANDS];
+#endif
 	struct ieee80211_channel *pCfg80211_Channels;
 	struct ieee80211_rate *pCfg80211_Rates;
 
