Index: mt76x2e-p4rev-113843/src/common/ee_flash.c
===================================================================
--- mt76x2e-p4rev-113843.orig/src/common/ee_flash.c
+++ mt76x2e-p4rev-113843/src/common/ee_flash.c
@@ -29,7 +29,7 @@
 #ifdef RTMP_FLASH_SUPPORT
 
 #include	"rt_config.h"
-
+#include    "eeflash.h"
 
 
 
Index: mt76x2e-p4rev-113843/src/common/eeprom.c
===================================================================
--- mt76x2e-p4rev-113843.orig/src/common/eeprom.c
+++ mt76x2e-p4rev-113843/src/common/eeprom.c
@@ -26,6 +26,7 @@
 	Name		Date			Modification logs
 */
 #include "rt_config.h"
+#include "eeflash.h"
 
 struct chip_map{
 	UINT32 ChipVersion;
Index: mt76x2e-p4rev-113843/src/include/eeflash.h
===================================================================
--- /dev/null
+++ mt76x2e-p4rev-113843/src/include/eeflash.h
@@ -0,0 +1,20 @@
+#ifdef RTMP_FLASH_SUPPORT
+static USHORT EE_FLASH_ID_LIST[]={
+
+#ifdef RT6352
+	0x6352,
+	0x7620,
+#endif /* RT6352 */
+
+
+
+#ifdef MT76x2
+#ifdef RTMP_MAC_PCI
+	0x7662,
+#endif /* RTMP_MAC_PCI */
+#endif /* MT76x0 */
+
+};
+
+#define EE_FLASH_ID_NUM  (sizeof(EE_FLASH_ID_LIST) / sizeof(USHORT))
+#endif /* RTMP_FLASH_SUPPORT */
Index: mt76x2e-p4rev-113843/src/include/eeprom.h
===================================================================
--- mt76x2e-p4rev-113843.orig/src/include/eeprom.h
+++ mt76x2e-p4rev-113843/src/include/eeprom.h
@@ -85,27 +85,6 @@
 #define EEPROM_DEFAULT_FILE_PATH                     "/etc_ro/wlan/RT2880_RT2820_AP_2T3R_V1_6.bin"
 #endif /* EEPROM_DEFAULT_FILE_PATH */
 
-#ifdef RTMP_FLASH_SUPPORT
-static USHORT EE_FLASH_ID_LIST[]={
-
-#ifdef RT6352
-	0x6352,
-	0x7620,
-#endif /* RT6352 */
-
-
-
-#ifdef MT76x2
-#ifdef RTMP_MAC_PCI
-	0x7662,
-#endif /* RTMP_MAC_PCI */
-#endif /* MT76x0 */
-
-};
-
-#define EE_FLASH_ID_NUM  (sizeof(EE_FLASH_ID_LIST) / sizeof(USHORT))
-#endif /* RTMP_FLASH_SUPPORT */
-
 /* For ioctl check usage */
 #define EEPROM_IS_PROGRAMMED		0x80
 
