Index: mt76x2e-p4rev-113843/src/os/linux/cfg80211/cfg80211.c
===================================================================
--- mt76x2e-p4rev-113843.orig/src/os/linux/cfg80211/cfg80211.c
+++ mt76x2e-p4rev-113843/src/os/linux/cfg80211/cfg80211.c
@@ -819,7 +819,7 @@ static int CFG80211_OpsStaGet(struct wip
 	const UINT8 *pMac, struct station_info *pSinfo)
 #endif
 {
-	void *pAd;
+	VOID *pAd;
 	CMD_RTPRIV_IOCTL_80211_STA StaInfo;
 
 	CFG80211DBG(RT_DEBUG_TRACE, ("80211> %s ==>\n", __FUNCTION__));
@@ -1679,10 +1679,10 @@ static int CFG80211_OpsSurveyGet(
 
 	/* return the information to upper layer */
 	pSurvey->channel = ((CFG80211_CB *)(SurveyInfo.pCfg80211))->pCfg80211_Channels;
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,5,0)) /* not sure version */
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4,0,0)) /* not sure version */
     pSurvey->filled = SURVEY_INFO_TIME_BUSY | SURVEY_INFO_EXT_BUSY;
-    pSurvey->time_busy = SurveyInfo.ChannelTimeBusy; /* unit: us */
-    pSurvey->time_ext_busy = SurveyInfo.ChannelTimeExtBusy;
+    pSurvey->channel_time_busy = SurveyInfo.ChannelTimeBusy; /* unit: us */
+    pSurvey->channel_time_ext_busy = SurveyInfo.ChannelTimeExtBusy;
 #else
 	pSurvey->filled = NL80211_SURVEY_INFO_CHANNEL_TIME_BUSY |
 						NL80211_SURVEY_INFO_CHANNEL_TIME_EXT_BUSY;
@@ -2247,7 +2247,7 @@ static int CFG80211_OpsStaDel(struct wip
 	const UINT8 *pMacAddr)
 #endif
 {
-	void *pAd;
+	VOID *pAd;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0))
 	const u8 *pMacAddr = params->mac;
 #endif
@@ -2293,7 +2293,7 @@ static int CFG80211_OpsStaChg(struct wip
 	const UINT8 *pMacAddr, struct station_parameters *params)
 #endif
 {
-	void *pAd;
+	VOID *pAd;
 	CFG80211_CB *p80211CB;
 
 	CFG80211DBG(RT_DEBUG_TRACE,
@@ -2340,7 +2340,7 @@ static struct net_device* CFG80211_OpsVi
 	char *name, enum nl80211_iftype Type, u32 *pFlags, struct vif_params *pParams)
 #endif /* LINUX_VERSION_CODE: 4.1.0 */
 {
-	void *pAd;
+	VOID *pAd;
 	CMD_RTPRIV_IOCTL_80211_VIF_SET vifInfo;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0))
 	PWIRELESS_DEV pDev = NULL;
@@ -2382,7 +2382,7 @@ static int CFG80211_OpsVirtualInfDel(
 #endif /* LINUX_VERSION_CODE: 3.6.0 */	
 	)
 {
-	void *pAd;
+	VOID *pAd;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,6,0))		
 	struct net_device *dev = NULL;
 	dev = pwdev->netdev;
@@ -2419,7 +2419,7 @@ static int CFG80211_start_p2p_device(
 	struct wiphy *pWiphy,
 	struct wireless_dev *wdev)
 {
-	void *pAd;
+	VOID *pAd;
 	struct net_device *dev = wdev->netdev;
 	CFG80211DBG(RT_DEBUG_OFF, ("80211> %s, %s [%d]==>\n", __FUNCTION__, dev->name, dev->ieee80211_ptr->iftype));
 	MAC80211_PAD_GET(pAd, pWiphy);
@@ -2430,7 +2430,7 @@ static void CFG80211_stop_p2p_device(
     struct wiphy           *pWiphy,
     struct wireless_dev    *wdev)
 {
-    void *pAd;
+    VOID *pAd;
     struct net_device *dev = wdev->netdev;
     CFG80211DBG(RT_DEBUG_OFF, ("80211> %s, %s [%d]==>\n", __FUNCTION__, dev->name, dev->ieee80211_ptr->iftype));
     MAC80211_PAD_GET(pAd, pWiphy);
