Index: mt7603e-4.0.1.0/rlt_wifi/os/linux/rt_profile.c
===================================================================
--- mt7603e-4.0.1.0.orig/rlt_wifi/os/linux/rt_profile.c
+++ mt7603e-4.0.1.0/rlt_wifi/os/linux/rt_profile.c
@@ -222,7 +222,7 @@ NDIS_STATUS	RTMPReadParametersHook(RTMP_
 		{
 #ifndef OS_ABL_SUPPORT
 			// TODO: need to roll back when convert into OSABL code
-				 fsize = (ULONG)srcf->f_dentry->d_inode->i_size;
+				 fsize = (ULONG)srcf->f_path.dentry->d_inode->i_size;
 				if (buf_size < (fsize + 1))
 					buf_size = fsize + 1;
 #endif /* OS_ABL_SUPPORT */
