Index: mt7603e-4.0.1.0/rlt_wifi/os/linux/rt_linux.c
===================================================================
--- mt7603e-4.0.1.0.orig/rlt_wifi/os/linux/rt_linux.c
+++ mt7603e-4.0.1.0/rlt_wifi/os/linux/rt_linux.c
@@ -853,8 +853,8 @@ void RtmpOSFileSeek(RTMP_OS_FD osfd, int
 int RtmpOSFileRead(RTMP_OS_FD osfd, char *pDataPtr, int readLen)
 {
 	/* The object must have a read method */
-	if (osfd->f_op && osfd->f_op->read) {
-		return osfd->f_op->read(osfd, pDataPtr, readLen, &osfd->f_pos);
+	if (osfd->f_op ) {
+		return vfs_read(osfd, pDataPtr, readLen, &osfd->f_pos);
 	} else {
 		DBGPRINT(RT_DEBUG_ERROR, ("no file read method\n"));
 		return -1;
@@ -864,7 +864,7 @@ int RtmpOSFileRead(RTMP_OS_FD osfd, char
 
 int RtmpOSFileWrite(RTMP_OS_FD osfd, char *pDataPtr, int writeLen)
 {
-	return osfd->f_op->write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
+	return vfs_write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
 }
 
 
@@ -1918,7 +1918,7 @@ VOID RtmpDrvAllMacPrint(
 				sprintf(msg, "%04x = %08x\n", macAddr, macValue);
 
 				/* write data to file */
-				file_w->f_op->write(file_w, msg, strlen(msg), &file_w->f_pos);
+				RtmpOSFileWrite(file_w, msg, strlen(msg));
 
 				printk("%s", msg);
 				macAddr += AddrStep;
@@ -1968,7 +1968,7 @@ VOID RtmpDrvAllE2PPrint(
 				sprintf(msg, "%08x = %04x\n", eepAddr, eepValue);
 
 				/* write data to file */
-				file_w->f_op->write(file_w, msg, strlen(msg), &file_w->f_pos);
+				RtmpOSFileWrite(file_w, msg, strlen(msg));
 
 				printk("%s", msg);
 				eepAddr += AddrStep;
@@ -2003,12 +2003,10 @@ VOID RtmpDrvAllRFPrint(
 			 ("-->2) %s: Error %ld opening %s\n", __FUNCTION__,
 			  -PTR_ERR(file_w), fileName));
 	} else {
-		if (file_w->f_op && file_w->f_op->write) {
 			file_w->f_pos = 0;
 			/* write data to file */
-			file_w->f_op->write(file_w, pBuf, BufLen, &file_w->f_pos);
-		}
-		filp_close(file_w, NULL);
+     		RtmpOSFileWrite(file_w, pBuf, BufLen);
+	    	filp_close(file_w, NULL);
 	}
 	set_fs(orig_fs);
 }
